### PR TITLE
[SPARK-45216][SQL] Fix non-deterministic seeded Dataset APIs

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
This PR fixes a bug regarding non-deterministic seeded Dataset functions.

If we run the following example the result is the expected equal 2 columns:
```
val c = rand()
df.select(c, c)

+--------------------------+--------------------------+
|rand(-4522010140232537566)|rand(-4522010140232537566)|
+--------------------------+--------------------------+
|        0.4520819282997137|        0.4520819282997137|
+--------------------------+--------------------------+
```

But if we run use other similar APIs their result is incorrect:
```
val r1 = random()
val r2 = uuid()
val r3 = shuffle(col("x"))
val x = df.select(r1, r1, r2, r2, r3, r3)

+------------------+------------------+--------------------+--------------------+----------+----------+
|            rand()|            rand()|              uuid()|              uuid()|shuffle(x)|shuffle(x)|
+------------------+------------------+--------------------+--------------------+----------+----------+
|0.7407604956381952|0.7957319451135009|e55bc4b0-74e6-4b0...|a587163b-d06b-4bb...| [1, 2, 3]| [2, 1, 3]|
+------------------+------------------+--------------------+--------------------+----------+----------+
```

This is because the current implementation of `rand()` passes a random seed to `Rand`, but other functions like `random()`, `uuid()` and `shuffle()` don’t. Later the `ResolveRandomSeed` rule is adds the necessary seeds but since the resolution rules don’t track expression object identities they can’t map an expression object 2 times to the same transformed object. I.e. in case of `random()` the `UnresolvedFunction("random", Seq.empty, ...)` object is transformed to 2 different `Rand(UnresolvedSeed)` objects and then 2 different random seeds are chosen.

This PR explicitely adds the seeds.

### Why are the changes needed?
To fix the above bug.

### Does this PR introduce _any_ user-facing change?
Yes, fixes the above bug.

### How was this patch tested?
Added new UT.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by cloud-fan — APPROVED** [review]
good catch!

**Comment by cloud-fan** [comment]
cc @HyukjinKwon

File: python/pyspark/sql/connect/functions.py [hunk 1]
```diff
@@ -388,7 +390,7 @@ def rand(seed: Optional[int] = None) -> Column:
     if seed is not None:
         return _invoke_function("rand", lit(seed))
     else:
-        return _invoke_function("rand")
+        return _invoke_function("rand", lit(random.randint(0, sys.maxsize)))
```
Comment (allisonwang-db) [inline python/pyspark/sql/connect/functions.py:393]: These are spark connect functions. Do we also need to update `python/pyspark/sql/functions.py`?

Comment (peter-toth) [reply python/pyspark/sql/connect/functions.py:393]: No, we don't. The sql Python functions just invoke the Scala implementations, which has been fixed.

BTW, I covered sql Python functions with a test: https://github.com/apache/spark/pull/42997/files#diff-a4984706554eae562eb0c7a66d284b1855d9bc463b03474f73493c7694244b66R1356-R1366

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by peter-toth** [comment]
Thanks for the review!

**Comment by JoshRosen** [comment]
I have a question about the scope of the bug fixed in this PR:

Is this issue specific to Connect? Or is there a pre-existing correctness issue that we might want to fix in other branches as well?

**Comment by HyukjinKwon** [comment]
This is a pre-existing correctness issue in DataFrame API with/without Spark Connect, that we might need to backport to other branches. I did not backport because it at least changes the output values but I don't mind we backport as it's technically a correctness issue.

**Pull request closed without merging**